### ISSUE TITLE
Problems with systemd socket activation: `Errno::EINVAL: Invalid argument - Invalid file`

### ISSUE THREAD

**Issue by paulmenzel** [puma/puma#976]
Using Puma 2.16.0 and JRuby 1.7.19 with systemd socket activation fails with 
the error `Errno::EINVAL: Invalid argument - Invalid file`.

```
$ sudo journalctl -u myapp-production.service
[…]
Mai 02 15:48:16 vsrv-myapp-2.comp.local systemd[1]: myapp-production.service holdoff time over, scheduling restart.
Mai 02 15:48:16 vsrv-myapp-2.comp.local systemd[1]: Started Puma application server (myapp production).
Mai 02 15:48:16 vsrv-myapp-2.comp.local systemd[1]: Starting Puma application server (myapp production)...
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: Errno::EINVAL: Invalid argument - Invalid file
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: for_fd at org/jruby/ext/socket/RubyBasicSocket.java:117
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: import_from_env at /srv/www/myapp/DashBoard/shared/bundle/jruby/1.9/gems/puma-2.16.0-java/lib/puma/binder.rb:61
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: times at org/jruby/RubyFixnum.java:280
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: import_from_env at /srv/www/myapp/DashBoard/shared/bundle/jruby/1.9/gems/puma-2.16.0-java/lib/puma/binder.rb:59
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: each at org/jruby/RubyHash.java:1341
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: import_from_env at /srv/www/myapp/DashBoard/shared/bundle/jruby/1.9/gems/puma-2.16.0-java/lib/puma/binder.rb:52
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: initialize at /srv/www/myapp/DashBoard/shared/bundle/jruby/1.9/gems/puma-2.16.0-java/lib/puma/cli.rb:55
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: (root) at /srv/www/myapp/DashBoard/shared/bundle/jruby/1.9/gems/puma-2.16.0-java/bin/puma:8
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: load at org/jruby/RubyKernel.java:1087
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: (root) at /home/capistrano/.rvm/gems/jruby-1.7.19@myapp/bin/puma:1
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: eval at org/jruby/RubyKernel.java:1107
Mai 02 15:48:23 vsrv-myapp-2.comp.local bootup_puma[3920]: (root) at /home/capistrano/.rvm/gems/jruby-1.7.19@myapp/bin/jruby_executable_hooks:15
Mai 02 15:48:23 vsrv-myapp-2.comp.local systemd[1]: myapp-production.service: main process exited, code=exited, status=1/FAILURE
Mai 02 15:48:23 vsrv-myapp-2.comp.local systemd[1]: Unit myapp-production.service entered failed state.
Mai 02 15:48:23 vsrv-myapp-2.comp.local systemd[1]: myapp-production.service failed.
```

Using no socket activation, or running the command used for `Exec` directly works successfully.

```
$ more /etc/systemd/system/myapp-production.{service,socket}
::::::::::::::
/etc/systemd/system/myapp-production.service
::::::::::::::
[Unit]
Description=Puma application server (myapp production)
After=network.target

Requires=myapp-production.socket

[Service]
Type=simple
WorkingDirectory=/srv/www/myapp/DashBoard/current
Environment=RAILS_ENV=production
User=capistrano
PIDFile=/srv/www/myapp/DashBoard/current/tmp/pids/puma.pid
ExecStart= /home/capistrano/.rvm/bin/bootup_puma \
            -e production -C /srv/www/myapp/DashBoard/shared/config/puma.rb

Restart=always

[Install]
WantedBy=multi-user.target
::::::::::::::
/etc/systemd/system/myapp-production.socket
::::::::::::::
[Unit]
Description=Puma HTTP Server Accept Sockets (myapp production)

[Socket]
#ListenStream=0.0.0.0:9292

# AF_UNIX domain socket
# SocketUser, SocketGroup, etc. may be needed for Unix domain sockets
SocketUser=capistrano
SocketGroup=capistrano
ListenStream=/srv/www/myapp/DashBoard/shared/tmp/sockets/puma.sock

# Socket options matching Puma defaults
NoDelay=true
ReusePort=true
Backlog=1024

[Install]
WantedBy=sockets.target
```

This is the used configuration file.

```
$ sudo more /srv/www/myapp/DashBoard/shared/config/puma.rb
#!/usr/bin/env puma

directory '/srv/www/myapp/DashBoard/current'
rackup "/srv/www/myapp/DashBoard/current/config.ru"

pidfile "/srv/www/myapp/DashBoard/shared/tmp/pids/puma.pid"
state_path "/srv/www/myapp/DashBoard/shared/tmp/pids/puma.state"
stdout_redirect '/srv/www/myapp/DashBoard/shared/log/puma_access.log', '/srv/www/myapp/DashBoard/shared/log/puma_error.log', true

threads 0,16

bind 'unix:///srv/www/myapp/DashBoard/shared/tmp/sockets/puma.sock?umask=0111'
workers 0

preload_app!
```

**Comment by dekellum** [comment]
Have you had a chance to try this under latest Puma 3.4.0? There were some fixes for socket activation that landed in 3.3.0.  Also latest jruby 1.7.25 or 9.x might help.   That said, I'm not sure how much socket activation has been tested for Puma on JRuby in particular.

**Comment by evanphx** [comment]
Please upgrade to 3.4.0 or later.
